content/source/assets/javascripts/application.js

```diff
@@ -7,9 +7,51 @@
 
 //= require analytics
 
+// Set up terraform.io UI helpers
 document.addEventListener("turbolinks:load", function() {
     "use strict";
+
+    // In navigation sidebars, hide most sub-lists and reveal any that contain the
+    // current page. The a.current-page class is added during build by
+    // layouts/inner.erb.
     var docsSidebar = $("#docs-sidebar ul.nav.docs-sidenav");
     docsSidebar.find("ul.nav").addClass("nav-hidden");
     docsSidebar.find("li").has("a.current-page").addClass("active");
+
+
+    // On docs/content pages, add a hierarchical quick nav menu if there are any
+    // H2/H3/H4 headers.
+    // Check whether turbolinks cached the quick-nav elements:
+    if ($("#inner #inner-quicknav ul").length == 0) {
+        // Build the quick-nav
+        var headers = $('#inner').find('h2, h3, h4');
+        if (headers.length > 0) {
+            $("#inner #inner-quicknav").html(
+                '<span id="inner-quicknav-trigger">Page Quick Nav<svg width="9" height="5" xmlns="http://www.w3.org/2000/svg"><path d="M8.811 1.067a.612.612 0 0 0 0-.884.655.655 0 0 0-.908 0L4.5 3.491 1.097.183a.655.655 0 0 0-.909 0 .615.615 0 0 0 0 .884l3.857 3.75a.655.655 0 0 0 .91 0l3.856-3.75z" fill-rule="evenodd"/></svg></span><ul class="dropdown"></ul>'
+            );
+            var quickNav = $('#inner-quicknav > ul.dropdown');
+            headers.each(function(index, element) {
+                var level = element.nodeName.toLowerCase();
+                var header_text = $(element).text();
+                var header_id = $(element).attr('id');
+                quickNav.append('<li class="level-' + level + '"><a href="#' + header_id + '">' + header_text + '</a></li>');
+            });
+        }
+    }
+    // Always attach event listeners, since turbolinks doesn't cache them
+    var quickNavTrigger = $('#inner-quicknav #inner-quicknav-trigger');
+    var quickNav = $('#inner-quicknav > ul.dropdown');
+    quickNavTrigger.on('click', function(e) {
+        $(this).siblings('ul').toggleClass('active');
+        e.stopPropagation();
+    });
+    quickNav.on('click', function(e) {
+        e.stopPropagation();
+    });
+    quickNav.find('li a').on('click', function() {
+        quickNav.removeClass('active');
+    });
+    $('body').on('click', function() {
+        quickNav.removeClass('active');
+    });
```

Comment:
I think this event handler is going to be added again and again for every page transition. You might want to namespace it, and remove it first? `$('body').off('click.quicknav'); $('body').on('click.quicknav', function() { …`

Reply:
Haha I had the same concern in my review above, but it seems as though if turbolinks replaces the entire body element and no references to the previous element remain, the event will be removed and garbage collected. It does still make me uncomfortable to see this though.

Reply:
I don't think turbolinks replaces the `body` element, just its contents. For event handlers on all other elements I think we're safe, but I think `body` is a special case. I could be mistaken though!

Reply:
Took a quick look in their docs and it says it does, although it might be worth verifying: https://github.com/turbolinks/turbolinks#navigating-with-turbolinks

Reply:
Ah, because it's on `<body>` instead of a sub-element? 

I tested this by having that handler just log something to the console, and navigated across a whole bunch of pages. But it just kept logging only once per click. So, I think that means it's not being re-added. 

But if you think it's good practice to do the off/on thing, I can do it anyway!

Reply:
lol sorry, I commented without reloading the page first. Well, anyway, I verified it. 😅